netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java

```diff
@@ -203,13 +221,35 @@ private void onDataRead(int streamId, ByteBuf data, boolean endOfStream) throws
   }
 
   /**
+   * Reports the appropriate {@link Status} for the stream.
+   */
+  private void reportStatus(NettyClientStream stream, @Nullable Status status,
+                            boolean stopDelivery) {
+    if (!stream.canReceive()) {
+      // The stream status has already been reported.
+      return;
+    }
+
+    // Just use the GOAWAY status if previously set.
+    status = goAwayStatus != null ? goAwayStatus : status;
+    if (status == null) {
+      // No cause was provided, no idea why we're shutting down.
+      status = Status.UNKNOWN.withDescription("Stream closing for unknown reason");
+    }
+
+    // Set the status on the stream if not already set.
+    stream.transportReportStatus(status, stopDelivery, new Metadata.Trailers());
+  }
+
+  /**
    * Handler for an inbound HTTP/2 RST_STREAM frame, terminating a stream.
    */
   private void onRstStreamRead(int streamId)
       throws Http2Exception {
     // TODO(nmittler): do something with errorCode?
-    NettyClientStream stream = clientStream(requireHttp2Stream(streamId));
-    stream.transportReportStatus(Status.UNKNOWN, false, new Metadata.Trailers());
+    // Only need to report status, the base handler will close the stream after returning
+    // from this method.
+    reportStatus(clientStream(requireHttp2Stream(streamId)), null, false);
```

Comment:
We were just using UKNOWN before ... are you asking to maintain that or to use a mapping from the HTTP/2 error code?